Sources/StreamChat/Utils/Database/NSManagedObject+Extensions.swift

```diff
@@ -10,13 +10,109 @@ extension NSManagedObject {
     }
 }
 
+extension NSEntityDescription {
+    class func insertNewObject<T: NSManagedObject>(
+        into context: NSManagedObjectContext,
+        for request: NSFetchRequest<T>
+    ) -> T {
+        let obj = T(context: context)
+        request.entity = obj.entity
+        FetchCache.shared.set(request, objectIds: [obj.objectID])
+        return obj
+    }
+}
+
+protocol NSFetchRequestGettable {}
+
+private let idKey = "id"
+
+extension NSFetchRequestGettable where Self: NSManagedObject {
+    static func fetchRequest(id: String) -> NSFetchRequest<Self> {
+        fetchRequest(keyPath: idKey, equalTo: id)
+    }
+    
+    static func fetchRequest(keyPath: String, equalTo value: String) -> NSFetchRequest<Self> {
+        let request = NSFetchRequest<Self>(entityName: entityName)
+        request.predicate = NSPredicate(format: "%K == %@", keyPath, value)
+        return request
+    }
+}
+
+extension NSManagedObject: NSFetchRequestGettable {}
+
 extension NSManagedObject {
-    static func load<T>(by request: NSFetchRequest<T>, context: NSManagedObjectContext) -> [T] {
+    static func load<T: NSManagedObject>(by id: String, context: NSManagedObjectContext) -> [T] {
+        load(keyPath: idKey, equalTo: id, context: context)
+    }
+    
+    static func load<T: NSManagedObject>(keyPath: String, equalTo value: String, context: NSManagedObjectContext) -> [T] {
+        let request = NSFetchRequest<T>(entityName: entityName)
+        request.predicate = NSPredicate(format: "%K == %@", keyPath, value)
+        return load(by: request, context: context)
+    }
+    
+    static func load<T: NSManagedObject>(by request: NSFetchRequest<T>, context: NSManagedObjectContext) -> [T] {
+        request.entity = NSEntityDescription.entity(forEntityName: T.entityName, in: context)!
         do {
-            return try context.fetch(request)
+            return try context.fetch(
+                request,
+                using: FetchCache.shared
+            )
         } catch {
             log.error("Could not load \(error), \(error.localizedDescription)")
             return []
         }
     }
 }
+
+class FetchCache {
+    fileprivate static let shared = FetchCache()
+    private let queue = DispatchQueue(label: "io.stream.com.fetch-cache", qos: .userInitiated)
```

Comment:
We can make this more performant if we make it concurrent:
```suggestion
    private let queue = DispatchQueue(label: "io.stream.com.fetch-cache", qos: .userInitiated, attributes: .concurrent)
```

Then, we make our write async, and protect it with a barrier flag:
```swift
       queue.async(flags: .barrier) {
            objectIDs = cache[request]
        }
```